### PR TITLE
Add test component with all ciphers and AEADs accelerated only

### LINKED ISSUE #6316: [G3] Add test component with all ciphers and AEADs accelerated only
- Add a test component that accelerates all ciphers, modes and AEADs, with as many things enabled as currently possible
- Add a reference component with a similar config and everything built-in, and coverage comparison in `analyze_outcome.py` as usual.

For this task, it is acceptable to disable as many modules as needed initially to make things build without built-in implementations of ciphers / modes / AEADs.  Adding support for new things is out of scope here, but small bugs can be fixed immediately if easy enough. It is also acceptable to ignore as much as necessary in the coverage comparison for now.

All gaps identified here will be either decided out-of-scope for this EPIC's goals, or closed by subsequent tasks to be created.

If bugs are fixed, the fixes should be backported if the bugs were also present in 2.28.

Depends on #8064

### PR DESCRIPTION
**Pull request by valeriosetti**
## Description

This PR resolves #6316 by adding a couple of test component in `all.sh`:
- one which accelerates all AEADs and ciphers (`component_test_psa_crypto_config_accel_cipher_aead`)
- one with builtin AEADs/ciphers (`component_test_psa_crypto_config_reference_cipher_aead`)

Both of them start from the `crypto` config in order to keep things at minimum working level.

A new task was also added to `analyze_outcomes.py` in order to ensure that test parity is achieved between the 2 scenarios.

~This PR depends on #8297~ (merged)

## PR checklist

- [x] **changelog** not required - test only
- [x] **backport** not required
- [x] **tests** provided: new test components added

### CODE REVIEW

File: tests/scripts/all.sh
Comment (mpg) [inline tests/scripts/all.sh:207]: Looks like leftovers from debugging :)

File: tests/scripts/all.sh
Comment (mpg) [inline tests/scripts/all.sh:3638]: Question: does something break if we change that to `crypto_full`? If yes, then let's create an issue about it, otherwise (or if it's very easy to fix) let's use `crypto_full`.

(Note: fully agree with the choice of excluding X.509 and TLS for starters, but when given a choice between `xxx` and `xxx_full` we should generally go for the full version if we can.)

File: tests/scripts/all.sh
Comment (mpg) [inline tests/scripts/all.sh:3678]: I don't think the `*` is wanted here (and if it were, it should be quoted to avoid it being interpreted by the shell).

(Same for next 3 lines.)

File: tests/scripts/all.sh
Comment (mpg) [inline tests/scripts/all.sh:3708]: I don't think PKCS5/12 is similar to `ALG_PBKDF2_AES_CMAC_PRF_128` though. I'd rather either:
- keep those enabled on both sides, but in `analyze_outcomes.py`, excluded tests that require a software cipher from the corresponding ciphersuites (including pkparse), or
- just disable them on both sides, if it turns out that close to nothing from these modules work without software ciphers.

(Note: the decision might be different for PKCS5 and PKCS12. I honestly don't know.)

File: tests/scripts/all.sh
Comment (mpg) [inline tests/scripts/all.sh:3697]: Question: why not go for a common `config_` function like we did for most other driver-reference pairs?

**Review by mpg — COMMENTED** [review]
Just an early look. This is clearly going in the right direction, but I still had a few questions and suggestions.

File: tests/scripts/all.sh
Comment (valeriosetti) [reply tests/scripts/all.sh:207]: Ops, sorry again for this :(

File: tests/scripts/all.sh
Comment (valeriosetti) [reply tests/scripts/all.sh]: It just took some guards' fix to make it work (because `full` means `USE_PSA_CRYPTO`), but it was pretty easy ;)

File: tests/scripts/all.sh
Comment (mpg) [reply tests/scripts/all.sh:207]: No worries, happens!

File: tests/scripts/analyze_outcomes.py
Comment (mpg) [inline tests/scripts/analyze_outcomes.py:296]: Note: out of scope (should be a separate PR), but I've been thinking for some time about adding the ability to ignore tests that match a regex. Here having just two lines for `PBES2 Decrypt` and `PBES2 Encrypt` would be more readable (and manageable, if we add more tests in the future).

File: tests/scripts/all.sh
Comment (mpg) [reply tests/scripts/all.sh:207]: Oops, you did it again ;)

File: tests/scripts/analyze_outcomes.py
Comment (mpg) [inline tests/scripts/analyze_outcomes.py:332]: Same: things would be much more readable & maintainable if we could ignore by pattern, here `PKCS#8 encrypted` for example.

File: tests/scripts/all.sh
Comment (mpg) [inline tests/scripts/all.sh:207]: ;)

**Review by mpg — CHANGES_REQUESTED** [review]
Thanks for addressing my comments! This is nearly an approval, except for the `-O0 -g` thing ;) (Btw, I hear `-Og -g` is supposed to be better.)

File: tests/scripts/analyze_outcomes.py
Comment (valeriosetti) [reply tests/scripts/analyze_outcomes.py:296]: I would totally agree with that. This morning, while adding these "skip lines" I realized that most of the lines in `analyze_outcomes.py` script are for test being ignored...

My initial idea of improvement was to just specify to ignore tests based on a list of dependencies, but this would mean that for each disparity the script has to go to the corresponding `*.data` file, look for the disparity test and check its dependencies. However the main drawback of this approach would be that we do not have a direct control over which tests are skipped in the coverage comparison and which not. I mean, it's pretty easy to just select a dependency that skips a lot of test both in the accelerated version and in the reference one without realizing how many tests are skipped.

Your proposal looks simpler to implement and perhaps give the same amount of control we have now. I will create a new task for this.

Comment (gilles-peskine-arm) [reply tests/scripts/analyze_outcomes.py:296]: > ignore tests based on a list of dependencies

I think it was a deliberate choice not to do that, because the goal was to catch undesired dependencies, typically dependencies that we forgot to update. It could be ok if there was one specific dependency `SKIPPED_IF_DRIVER_ONLY`.

I think a regex is easier though. We do have to be disciplined enough to make the regexes both specific enough (to not catch things inadvertently) and readable (so we can easily tell if something is on the ignore list).

Comment (mpg) [reply tests/scripts/analyze_outcomes.py:296]: Just created #8365 about that, and when coming back to add a link, I found out I had missed the last two comments here.

Comment (mpg) [reply tests/scripts/analyze_outcomes.py:296]: > I think it was a deliberate choice not to do that, because the goal was to catch undesired dependencies, typically dependencies that we forgot to update. It could be ok if there was one specific dependency `SKIPPED_IF_DRIVER_ONLY`.

Indeed it was a deliberate choice, for this reason.

Comment (valeriosetti) [reply tests/scripts/analyze_outcomes.py:296]: @mpg is it me or #8365 is pretty similar to #8361?

Comment (mpg) [reply tests/scripts/analyze_outcomes.py:296]: Yeah, sorry, I failed to notice you had already created 8361 when I created 8365, otherwise I'd have added my thoughts to 8361 as comments.

Comment (valeriosetti) [reply tests/scripts/analyze_outcomes.py:296]: No problem at all, I was only asking to understand if it was worth to keep #8361 around or not. Since it's not, then I will close it as duplicate

**Comment by mpg** [comment]
@valeriosetti Can you rebase this now that #8297 has been merged? Then I'll review again tomorrow.

File: tests/suites/test_suite_cipher.function
Comment (mpg) [inline tests/suites/test_suite_cipher.function:186]: Minor: I think using `TEST_ASSUME()` would be more precise here - it will mark the test as skipped rather than passing when the assumption doesn't hold.

**Review by mpg — CHANGES_REQUESTED** [review]
Almost all good, just one minor thing.

**Review by mpg — APPROVED** [review]
LGTM!

**Review by mpg — APPROVED** [review]
LGTM

File: tests/scripts/analyze_outcomes.py
Comment (mpg) [inline tests/scripts/analyze_outcomes.py:271]: On second thought, this part (`test_suite_psa_crypto`) is not really satisfying: do these tests really need the built-in implementations? If so, the reason should be explained here. If not, the dependency declarations should be fixed. (Could be a mix of both depending on why test, or course.)

**Review by mpg — CHANGES_REQUESTED** [review]
Sorry, changed my mind when I re-read this while working on #8365.

File: tests/scripts/analyze_outcomes.py
Comment (gilles-peskine-arm) [reply tests/scripts/analyze_outcomes.py:271]: None of the test cases here under `test_suite_psa_crypto` should depend on the built-in implementation.

I have not checked whether it makes more sense to add an exception now and fix them later, or fix them now.

**Review by mpg — APPROVED** [review]
LGTM if the CI agrees.

File: tests/scripts/all.sh
Comment (mpg) [inline tests/scripts/all.sh:3601]: Actually, after #8314 I think it would be better to accelerate CMAC than to disable it. (In the general spirit of disabling on the minimum.) So:
- don't unset `ALG_CMAC` or PBKDF2-CMAC  here but add `ALG_CMAC` to `loc_accel_list` in the accel component;
- don't unset `CMAC_C` here but only in the accel component;
- ignore `test_suite_cmac` in `analyze_outcomes.py`.

(This will need rebasing on recent development.)

File: tests/scripts/all.sh
Comment (mpg) [inline tests/scripts/all.sh:3602]: Similarly, after #8315 I don't think we should be disabling STREAM or ECB, but rather declare them as accelerated.

**Review by mpg — CHANGES_REQUESTED** [review]
Sorry, again I found a few more things to complain about while looking at the code while planning for future tasks.

**Comment by valeriosetti** [comment]
I had to rebase the PR in order to include what it has been added from #8314 and #8315 in order to address these comments:
- https://github.com/Mbed-TLS/mbedtls/pull/8303#discussion_r1371425816
- https://github.com/Mbed-TLS/mbedtls/pull/8303#discussion_r1371427278

**Review by mpg — APPROVED** [review]
LGTM if the CI agrees!

File: library/cipher.c
Comment (gilles-peskine-arm) [inline library/cipher.c:269]: This is getting increasingly hard to read.

```suggestion
    if (mbedtls_cipher_get_base(cipher_info)->ctx_alloc_func != NULL) {
        ctx->cipher_ctx = mbedtls_cipher_get_base(cipher_info)->ctx_alloc_func();
        if (ctx->cipher_ctx == NULL) {
            return MBEDTLS_ERR_CIPHER_ALLOC_FAILED;
        }
    }
```

File: tests/scripts/analyze_outcomes.py
Comment (gilles-peskine-arm) [inline tests/scripts/analyze_outcomes.py:222]: This comment doesn't make sense to me: we don't accelerate suites. I guess the intention is to ignore suites that are specific to the built-in implementation of mechanisms that are accelerated in the driver component?

File: tests/scripts/analyze_outcomes.py
Comment (gilles-peskine-arm) [inline tests/scripts/analyze_outcomes.py:257]: The tests don't. It's the PEM module that depends on the built-in algorithms. Same with pkcs5 and pkcs12 below.

File: tests/scripts/analyze_outcomes.py
Comment (gilles-peskine-arm) [inline tests/scripts/analyze_outcomes.py:278]: More to the point, they depend on the built-in implementation of CBC-PKCS7.

Same for pkcs5 and pkparse.

File: tests/scripts/analyze_outcomes.py
Comment (gilles-peskine-arm) [inline tests/scripts/analyze_outcomes.py:229]: Looks like we should add the ability to list `'aes'` and have all `'aes.xxx'` ignored. Actually, don't we already have this ability? Can't we replace these lines with just `'aes'`, and so on below?

File: tests/scripts/analyze_outcomes.py
Comment (gilles-peskine-arm) [inline tests/scripts/analyze_outcomes.py:259]: Not a blocker, but we'll need a follow-up cleanup when both #8398 and #8303 are merged.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
LGTM except some readability issues.

File: tests/scripts/analyze_outcomes.py
Comment (valeriosetti) [reply tests/scripts/analyze_outcomes.py:222]: Sorry, that was the idea but I phrased it very badly

File: tests/scripts/analyze_outcomes.py
Comment (mpg) [reply tests/scripts/analyze_outcomes.py:259]: Yes, my intention was to update 8398 if this one gets merged first. I even had a local commit doing that: https://github.com/Mbed-TLS/mbedtls/commit/4207c655e02add507a5510a88261dbc5ea7293b0 - just waiting for this one to get merged before applying it. I'm labeling 8398 as "needs-preceding-pr" as a reminder.

File: tests/scripts/analyze_outcomes.py
Comment (mpg) [reply tests/scripts/analyze_outcomes.py:257]: Agreed. I changed this (and a few other similar comments) in my local copy of 8389+this: https://github.com/Mbed-TLS/mbedtls/commit/4207c655e02add507a5510a88261dbc5ea7293b0#diff-f476aaedf3af73a4f55d7527600c29840f3b7cce70fa534b988c375e34f93dfbR278

File: tests/scripts/analyze_outcomes.py
Comment (mpg) [reply tests/scripts/analyze_outcomes.py:229]: Ah, good point, I completely forgot we already had this ability even though I re-read that part of the code recently.

File: tests/scripts/analyze_outcomes.py
Comment (valeriosetti) [reply tests/scripts/analyze_outcomes.py:257]: Ok, so I will use the same sentences so that you can skip that commit ;)

File: tests/scripts/analyze_outcomes.py
Comment (mpg) [reply tests/scripts/analyze_outcomes.py:222]: FWIW, in #8398 I went with "Modules replaced by drivers".

**Review by mpg — APPROVED** [review]
LGTM, thanks!

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM

**Pull request merged**
